snowdrop-block/src/Snowdrop/Block/State.hs

```diff
@@ -87,76 +78,45 @@ mapTxs_ handleTxDo txsWithAccs =
-    , HUpCastableChSet '[BlundComponent blkType] xs
+    , QueryERo xs (TipComponent blkType)
     , UpCastableERoM (UnionSeqExpandersInps txtypes) xs
-    , Default (HChangeSet xs)
-    , HasGetter (Undo conf) (BlockUndo blkType)
-    , HasReview (Undo conf) (BlockUndo blkType)
     )
     => BlkConfiguration blkType
     -> Validator conf txtypes
     -> Rec (ProofNExp conf) txtypes
-    -> (RawBlk blkType -> [SomeTx c] -> Block (BlockHeader blkType) [SomeTx c])
-    -> BlkStateConfiguration blkType m
-inmemoryBlkStateConfiguration cfg validator exps mkBlock = fix $ \this ->
+    -> BlkStateConfiguration (ChgAccum conf) blkType m
+inmemoryBlkStateConfiguration cfg validator exps =
     BlkStateConfiguration {
-      bscConfig = cfg
-    , bscExpand = \rawBlock -> do
-          blkPayload <-
-              liftERoComp $ upcastEffERoCompM @_ @xs $
-                  runSeqExpandersSequentially exps (gett rawBlock)
-          pure (mkBlock rawBlock blkPayload)
-    , bscApplyPayload = \(gett @_ @[SomeTx (BlkProcConstr txtypes xs)] -> txs) -> do
-          (newSt, undo) <- liftERoComp $ do
-              curAcc <- asks (getCAOrDefault @conf . gett)
-              OldestFirst accs <-
-                  convertEffect @conf $ modifyAccum @xs @conf (OldestFirst $ map (applySomeTx $ hupcast . txBody) txs)
-              let preAccs = init (curAcc :| accs)
-                  lastAcc = last (curAcc :| accs)
-              convertEffect @conf $ mapTxs_ @conf (runValidator validator) (zip preAccs txs)
-              (lastAcc,) . gett <$> computeUndo @xs @conf lastAcc
-          undo <$ modify (flip sett newSt)
-     , bscApplyUndo = liftERoComp . modifyAccumUndo @xs @conf . pure . inj >=> modify . flip sett
-     , bscStoreBlund = \blund -> do
-           let blockRef = unCurrentBlockRef $ bcBlockRef cfg (blkHeader $ buBlock blund)
-           let chg = hChangeSetFromMap @(BlundComponent blkType) $ M.singleton blockRef (New blund)
-           applyBlkChg chg
-     , bscRemoveBlund = \blockRef -> do
-           let chg = hChangeSetFromMap @(BlundComponent blkType) $ M.singleton blockRef Rem
-           applyBlkChg chg
-     , bscGetBlund = liftERoComp . queryOne @(BlundComponent blkType) @xs @conf
+      bscVerifyConfig = cfg
+    , bscExpandHeaders = \_ rawBlocks -> do
+      pure $ OldestFirst $ gett <$> rawBlocks
+    , bscExpandPayloads = \chgAccum rawBlks -> let
+        comp = OldestFirst <$> traverse (runSeqExpandersSequentially exps . gett) rawBlks
+        in liftERoComp $ upcastEffERoCompM @_ @xs $ local (lensFor @(Ctx conf) @(ChgAccumCtx conf) .~ CAInitialized @conf chgAccum) comp
+    , bscValidateTx = \chgAccum tx -> liftERoComp $ mapTx_ @conf (runValidator validator) (chgAccum, tx)
+    , bscGetHeader = \blockRef -> liftERoComp $ do
+        blund <- queryOne @(BlundComponent blkType) @xs @conf $ blockRef
+        pure $ blkHeader. buBlock <$> blund
+    , bscInmemRollback = \blockRef -> liftERoComp $ do
```

Comment:
Implementation is wrong: it takes the blund with given `blockRef` and rolls it back. While the desired behavior is to sequentially rollback every block from the tip down to `blockRef`.

Reply:
So the proper implementation would be to sequentially query blocks from `blockRef` to tip, obtain list `blunds`.
Then use operation `modifyAccumUndo $ buUndo <$> blunds` instead of `modifyAccumOne @_ @conf $ gett @_ @(HChangeSet xs) blockUndo`